### PR TITLE
Show QR Code scanner as push view, not modal (#782)

### PR DESCRIPTION
**Pull request by BloodyPixy**
@vikmeup , referencing https://github.com/TrustWallet/trust-wallet-ios/issues/782 issue. I have started to alter the code. I have moved logic for creating and navigating QR Code scanner from `SendViewController` to `SendCoordinator`, since it seems to me, it is coordinators responsibility, and then having a controller reference, I update it using a new dedicated function.

I will proceed to implement this change in all the other places. If you have any comments, feel free.

Edit: I could not test a browser, because it is not working right now.

### CODE REVIEW

**Comment by OlegGordiichuk** [comment]
@BloodyPixy could you pls fix lint warnings.

**Comment by BloodyPixy** [comment]
@OlegGordiichuk , I am not done yet.

**Comment by BloodyPixy** [comment]
@OlegGordiichuk , @vikmeup , I have more or less finalized this. In some cases, this is impossible to handle via coordinator, as there in no any.

**Comment by vikmeup** [comment]
@BloodyPixy You gotta build a generic solution to push coordinator, I can work on this later, but there is an examples how it's done

**Comment by vikmeup** [comment]
@BloodyPixy take a look at this: https://irace.me/navigation-coordinators, we only need logic for `pushCoordinator()` make it as protocol?

**Comment by BloodyPixy** [comment]
@vikmeup , check new commits. I believe that's much better and can be applied to all coordinators.

File: Trust/UI/NavigationController.swift [hunk 1]
```diff
@@ -35,3 +37,26 @@ class NavigationController: UINavigationController {
...
+        viewControllersToChildCoordinators[coordinator.rootViewController] = coordinator
+        pushViewController(coordinator.rootViewController, animated: true)
+    }
+    func popCoordinator(_ coordinator: RootCoordinator) {
+        viewControllersToChildCoordinators.removeValue(forKey: coordinator.rootViewController)
+        popViewController(animated: true)
+    }
+}
+
+protocol RootViewControllerProvider: class {
```
Comment (vikmeup) [inline Trust/UI/NavigationController.swift]: can you move into separate file?

**Comment by vikmeup** [comment]
@BloodyPixy Nice work! I will take another look tomorrow!

File: Trust/Tokens/ViewControllers/NewTokenViewController.swift [hunk 2]
```diff
@@ -166,16 +168,14 @@ class NewTokenViewController: FormViewController {
...
 }
 
-extension NewTokenViewController: QRCodeReaderDelegate {
-    func readerDidCancel(_ reader: QRCodeReaderViewController!) {
-        reader.stopScanning()
-        reader.dismiss(animated: true, completion: nil)
+extension NewTokenViewController: ScanQRCodeCoordinatorDelegate {
+    func didCancel(in coordinator: ScanQRCodeCoordinator) {
+        navigationController?.setNavigationBarHidden(false, animated: false)
+        (navigationController as? NavigationController)?.popCoordinator(coordinator)
```
Comment (vikmeup) [inline Trust/Tokens/ViewControllers/NewTokenViewController.swift]: navigationController should be `NavigationController` in here?

File: Trust/UI/NavigationController.swift (hunk 1 above)
Comment (BloodyPixy) [reply Trust/UI/NavigationController.swift]: Even in two

File: Trust/Tokens/ViewControllers/NewTokenViewController.swift (hunk 2 above)
Comment (BloodyPixy) [reply Trust/Tokens/ViewControllers/NewTokenViewController.swift]: @vikmeup I know what you are talking about, but in our case, we don't always have a coordinator. Take a look on how NewTokenViewController & ImportWalletViewController are created. They don't have coordinators, but somehow, they need to present the QR Scanner. So it seems to me as a viable solution. Please, let me know if I am missing something.

Comment (vikmeup) [reply Trust/Tokens/ViewControllers/NewTokenViewController.swift]: View controllers can call delegate to the coordinator and you push from the coorditor then, that would be the right way to do it!

Comment (BloodyPixy) [reply Trust/Tokens/ViewControllers/NewTokenViewController.swift]: I agree. On it.

Comment (vikmeup) [reply Trust/Tokens/ViewControllers/NewTokenViewController.swift]: Awesome, this is gonna be a big improvement to our codebase once it’s done

File: Trust/Browser/Coordinators/BrowserCoordinator.swift [hunk 3]
```diff
@@ -41,6 +41,9 @@ class BrowserCoordinator: NSObject, Coordinator {
         controller.delegate = self
         return controller
     }()
+    var providedRootController: UIViewController {
```
Comment (vikmeup) [inline Trust/Browser/Coordinators/BrowserCoordinator.swift]: why `providedRootController`? just call it `rootViewController`?

File: Trust/Foundation/PushableCoordinator.swift [hunk 4]
```diff
@@ -1,10 +1,44 @@
...
+                return
+            }
+            addCoordinator(coordinator)
+            coordinator.providedRootController.hidesBottomBarWhenPushed = tabBarHidden
+            nvc.setNavigationBarHidden(navigationBarHidden, animated: false)
+            nvc.pushViewController(coordinator.providedRootController, animated: true)
+        }
+    }
+    func popCoordinator(_ coordinator: RootCoordinator, navigationBarHidden: Bool = false) {
+        if let presentedNVC = providedRootController.presentedViewController as? UINavigationController {
```
Comment (vikmeup) [inline Trust/Foundation/PushableCoordinator.swift]: why do you need to pop? it should be done automatically once view is closed

File: Trust/Browser/Coordinators/BrowserCoordinator.swift (hunk 3 above)
Comment (BloodyPixy) [reply Trust/Browser/Coordinators/BrowserCoordinator.swift]: +, it intersected with the existing name of controllers in some places, but I will just change them as well.

**Comment by OlegGordiichuk** [comment]
@BloodyPixy could you pls resolve conflicts?

**Comment by BloodyPixy** [comment]
Closing this PR, as there was a lot of discussion and changes involved. Creating new branch for this issue.

**Pull request closed without merging**